granulate_utils/linux/ns.py

```diff
@@ -62,33 +69,43 @@ def run_in_ns(nstypes: List[str], callback: Callable[[], T], target_pid: int = 1
     # make sure "mnt" is last, once we change it our /proc is gone
     nstypes = sorted(nstypes, key=lambda ns: 1 if ns == "mnt" else 0)
 
-    ret: Optional[T] = None
+    ret: Union[T, _Sentinel] = _SENTINEL
+    exc: Optional[BaseException] = None
 
     def _switch_and_run():
-        libc = ctypes.CDLL("libc.so.6")
-        for nstype in nstypes:
-            if not is_same_ns(target_pid, nstype):
-                flag = {
-                    "mnt": 0x00020000,  # CLONE_NEWNS
-                    "net": 0x40000000,  # CLONE_NEWNET
-                    "pid": 0x20000000,  # CLONE_NEWPID
-                    "uts": 0x04000000,  # CLONE_NEWUTS
-                }[nstype]
-                if libc.unshare(flag) != 0:
-                    raise ValueError(f"Failed to unshare({nstype})")
-
-                with open(f"/proc/{target_pid}/ns/{nstype}", "r") as nsf:
-                    if libc.setns(nsf.fileno(), flag) != 0:
-                        raise ValueError(f"Failed to setns({nstype}) (to pid {target_pid})")
-
-        nonlocal ret
-        ret = callback()
+        try:
+            libc = ctypes.CDLL("libc.so.6")
+            for nstype in nstypes:
+                if not is_same_ns(target_pid, nstype):
+                    flag = {
+                        "mnt": 0x00020000,  # CLONE_NEWNS
+                        "net": 0x40000000,  # CLONE_NEWNET
+                        "pid": 0x20000000,  # CLONE_NEWPID
+                        "uts": 0x04000000,  # CLONE_NEWUTS
+                    }[nstype]
+                    if libc.unshare(flag) != 0:
+                        raise ValueError(f"Failed to unshare({nstype})")
+
+                    with open(f"/proc/{target_pid}/ns/{nstype}", "r") as nsf:
+                        if libc.setns(nsf.fileno(), flag) != 0:
+                            raise ValueError(f"Failed to setns({nstype}) (to pid {target_pid})")
+
+            nonlocal ret
+            ret = callback()
+        except BaseException as e:
+            nonlocal exc
+            exc = e
```

Comment:
Explain in a comment why you're doing that (i.e., that you want the exception to take place in the calling thread)

Reply:
kk